internal/repository/in_memory_user_repository.go

```diff
@@ -0,0 +1,117 @@
+package repository
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ceciivanov/platform-go-challenge/internal/models"
+	"github.com/ceciivanov/platform-go-challenge/internal/repository/mock_data"
+)
+
+// InMemoryUserRepository is an in-memory implementation of the UserRepository interface
+// InMemoryUserRepository contains a map of all Users with their favorite assets
+type InMemoryUserRepository struct {
+	Users map[int]models.User
+	mu    sync.RWMutex // mu is a read-write mutex to protect the Users map from concurrent access
+}
+
+// NewInMemoryUserRepository creates a new instance of InMemoryUserRepository
+func NewInMemoryUserRepository() *InMemoryUserRepository {
```

Comment:
This is a great approach. Could you explain if it would be beneficial for horizontal scaling?

Reply:
By trying to implement horizontal scaling with this approach I think that we will have problem with data inconsistency because each instance of application will have it's own in memory data and this could lead to synchronisation challenges & issues, for example user's favorites may differ across different instances. 


To achieve horizontal scaling a good approach will be following service-oriented architecture where the application will break to micro services each handling different domain and will be independent of any other.